src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/AsyncMethodBuilderCore.cs

```diff
@@ -105,7 +105,7 @@ internal static string GetAsyncStateMachineDescription(IAsyncStateMachine stateM
         [MethodImpl(MethodImplOptions.NoInlining)]
         internal static void LogTraceOperationBegin(Task t, Type stateMachineType)
         {
-            TplEventSource.Log.TraceOperationBegin(t.Id, "Async: " + stateMachineType.Name, 0);
+            if (EventSource.IsSupported) TplEventSource.Log.TraceOperationBegin(t.Id, "Async: " + stateMachineType.Name, 0);
```

Comment:
The call to this method is under `if (TplEventSource.Log.IsEnabled())` condition. Why do we need to duplicate the condition here?

Reply:
(Dtto for all other places that have this duplication.)

Reply:
I think this should be done by telling the trimmer to replace `EventSource.IsEnabled()` with false when EventSource is not supported. It does not make sense to duplicate the `EventSource.IsSupported` condition everywhere.

Reply:
Yeah, I'm unhappy about size of that diff too. 
I need to investigate more, this is "stash it for later, it's too late tonight" . 

I suspect that `TplEventSource.Log.IsEnabled()` is holding `TplEventSource.Log` alive even after trimming.

Maybe I could introduce `TplEventSource.IsEnabled()` instead which would call `TplEventSource.Log.IsEnabled()` and trim/substitute that for `false`. But it must not trim when `EventSource.IsSupported` is true. And so it must be XML and not an attribute.

Reply:
> TplEventSource.Log alive even after trimming.

It is expected to be only a small dummy artifact. All EventSource methods that do any actual work have `if (IsSupported) return;` or similar, so all heavy lifting should be trimmed already.

Reply:
> Maybe I could introduce TplEventSource.IsEnabled()

I would focus on fixes in the EventSource itself. I would try to avoid changing every actual event source out there, even if it means leaving a tiny bit of size on the table.

Reply:
The trimmer is not that smart. Any mention of some class keeps the class alive and the class bodies are usually not trimmed. For example `DiagLinkedList` is static of `Activity` class. And ball of dead code keeps rolling.

Reply:
`activity1` is always null after trimming. 

```
      Activity activity1 = (Activity) null;
      if (!SocketsTelemetry.IsActivitySourceSupported)
        ;
      if (activity1 != null)
      {
               ... bloat
      }
```

![image](https://github.com/user-attachments/assets/76f77393-734b-4b7c-a94f-1702be2df4ef)

@sbomer is this bug that would be considered for fixing, or I should seek workarounds ?

Reply:
Created https://github.com/dotnet/runtime/issues/114730 with simple repro

Reply:
And maybe it's same as https://github.com/dotnet/linker/issues/1284